scalalib/src/scalafmt/ScalafmtWorker.scala

```diff
@@ -21,44 +22,84 @@ private[scalafmt] class ScalafmtWorker {
 
   def reformat(input: Seq[PathRef],
                scalafmtConfig: PathRef)(implicit ctx: Ctx): Unit = {
-    val toFormat =
-      if (scalafmtConfig.sig != configSig) input
-      else
-        input.filterNot(ref => reformatted.get(ref.path).contains(ref.sig))
-
-    if (toFormat.nonEmpty) {
-      ctx.log.info(s"Formatting ${toFormat.size} Scala sources")
-      reformatAction(toFormat.map(_.path),
-                     scalafmtConfig.path)
-      reformatted ++= toFormat.map { ref =>
-        val updRef = PathRef(ref.path)
-        updRef.path -> updRef.sig
-      }
-      configSig = scalafmtConfig.sig
+    reformatAction(input, scalafmtConfig, dryRun = false)
+  }
+
+  def checkFormat(input: Seq[PathRef],
+                  scalafmtConfig: PathRef)(implicit ctx: Ctx): Result[Unit] = {
+
+    val misformatted = reformatAction(input, scalafmtConfig, dryRun = true)
+    if (misformatted.isEmpty) {
+      Result.Success(())
     } else {
-      ctx.log.info(s"Everything is formatted already")
+      val out = ctx.log.outputStream
+      for (u <- misformatted) {
+        out.println(u.path.toString)
+      }
+      Result.Failure(s"Found ${misformatted.length} misformatted files")
     }
   }
 
-  private val cliFlags = Seq("--non-interactive", "--quiet")
+  // run scalafmt over input files and return any files that changed
+  // (only save changes to files if dryRun is false)
+  private def reformatAction(
+    input: Seq[PathRef],
+    scalafmtConfig: PathRef,
+    dryRun: Boolean
+  )(implicit ctx: Ctx): Seq[PathRef] = {
+
+    // only consider files that have changed since last reformat
+    val toConsider =
+      if (scalafmtConfig.sig != configSig) input
+      else input.filterNot(ref => reformatted.get(ref.path).contains(ref.sig))
+
+    if (toConsider.nonEmpty) {
 
-  private def reformatAction(toFormat: Seq[os.Path],
-                             config: os.Path)(implicit ctx: Ctx) = {
-    val scalafmt =
-      Scalafmt
+      if (dryRun) {
+        ctx.log.info(s"Checking format of ${toConsider.size} Scala sources")
+      } else {
+        ctx.log.info(s"Formatting ${toConsider.size} Scala sources")
+      }
+
+      val scalafmt = Scalafmt
         .create(this.getClass.getClassLoader)
         .withRespectVersion(false)
 
-    val configPath =
-      if (os.exists(config))
-        config.toNIO
-      else
-        JPaths.get(getClass.getResource("default.scalafmt.conf").toURI)
+      val configPath =
+        if (os.exists(scalafmtConfig.path))
+          scalafmtConfig.path.toNIO
+        else
+          JPaths.get(getClass.getResource("default.scalafmt.conf").toURI)
+
+      // keeps track of files that are misformatted
+      val misformatted = mutable.ListBuffer.empty[PathRef]
+
+      def markFormatted(path: PathRef) = {
```

Comment:
nit: I'd invoke `markFormatted` from `checkFormat` and `reformat` instead of in the `foreach` loop.